src/core/matchers/matchersUtil.js

```diff
@@ -168,25 +205,30 @@ getJasmineRequireObj().matchersUtil = function(j$) {
     if (className == '[object Array]') {
       size = a.length;
       if (size !== b.length) {
+        diffBuilder.record(a, b);
         return false;
       }
 
-      while (size--) {
-        result = eq(a[size], b[size], aStack, bStack, customTesters);
-        if (!result) {
-          return false;
-        }
+      for (i = 0; i < size; i++) {
+        diffBuilder.withPath(i, function() {
+          result = eq(a[i], b[i], aStack, bStack, customTesters, diffBuilder) && result;
+        });
+      }
+      if (!result) {
+        return false;
       }
     } else if (className == '[object Set]') {
       if (a.size != b.size) {
+        diffBuilder.record(a, b);
         return false;
       }
       var iterA = a.values(), iterB = b.values();
       var valA, valB;
       do {
         valA = iterA.next();
         valB = iterB.next();
-        if (!eq(valA.value, valB.value, aStack, bStack, customTesters)) {
+        if (!eq(valA.value, valB.value, aStack, bStack, customTesters, j$.NullDiffBuilder())) {
```

Comment:
Shouldn't this be the `diffBuilder` local and not grabbing a new `NullDiffBuilder`?